test/boost/memtable_test.cc

```diff
@@ -814,3 +815,90 @@ SEASTAR_TEST_CASE(failed_flush_prevents_writes) {
 #endif
 }
 
+SEASTAR_TEST_CASE(flushing_rate_is_reduced_if_compaction_doesnt_keep_up) {
+    // The test simulates a situation where N threads issue flushes to N
+    // tables. Half are issuing lots of small flushes and the other half
+    // issues a few large ones. This can lead to a situation where lots of
+    // small sstables accumulate on disk, and, if compaction never has a chance
+    // to keep up, resources can be exhausted.
+    return do_with_cql_env([](cql_test_env& env) -> future<> {
+        using result_t = std::tuple<unsigned, unsigned, unsigned>;
+        struct flusher {
+            cql_test_env& env;
+            const int num_flushes;
+            const int num_mutations_per_flush;
+            const int num_concurrent_flushes;
+            const int mutations_per_commitlog_sync = 50;
+
+            static sstring cf_name(unsigned thread_id) {
+                return format("cf_{}", thread_id);
+            }
+
+            static sstring ks_name() {
+                return "ks";
+            }
+
+            future<> create_table(schema_ptr s) {
+                service::migration_manager& mm = env.migration_manager().local();
+                auto group0_guard = co_await mm.start_group0_operation();
+                auto ts = group0_guard.write_timestamp();
+                auto announcement = co_await mm.prepare_new_column_family_announcement(s, ts);
+                co_await mm.announce(std::move(announcement), std::move(group0_guard));
+            }
+
+            future<result_t> operator()() {
+                const sstring ks_name = this->ks_name();
+                const sstring cf_name = this->cf_name(this_shard_id());
+                random_mutation_generator gen{
+                    random_mutation_generator::generate_counters::no,
+                    local_shard_only::yes,
+                    random_mutation_generator::generate_uncompactable::no,
+                    std::nullopt,
+                    ks_name.c_str(),
+                    cf_name.c_str()
+                };
+                schema_ptr s = gen.schema();
+
+                co_await create_table(s);
+                replica::database& db = env.local_db();
+                replica::table& t = db.find_column_family(ks_name, cf_name);
+
+                size_t max_sstables_count = 0;
+                semaphore sem{num_concurrent_flushes};
+                unsigned mutations_applied = 0;
+                auto range = boost::irange<int>(0, num_flushes);
+                co_await seastar::parallel_for_each(range.begin(), range.end(), [&](int value) -> future<> {
+                    auto permit = co_await get_units(sem, 1);
+                    std::vector<mutation> mutations = gen(num_mutations_per_flush);
+                    for (auto&& m : mutations) {
```

Comment:
Why more than one mutation?

Reply:
I don't think the generator api allows creation of one massive mutation, if you need a big change you need to produce a number of mutations. Regardless of that, the main reason is that it takes long enough time to generate (without causing reactor stalls) that it blocks the compaction of the smaller sstables (on the other shard) due to history update. It effectively simulates the scenario of compaction being so slow that it can't keep up with flushes.